core/trino-spi/src/main/java/io/trino/spi/type/Int128.java

```diff
@@ -0,0 +1,185 @@
+    private final long low;
+
+    public Int128(long high, long low)
+    {
+        this.high = high;
+        this.low = low;
+    }
+
+    public static Int128 valueOf(int[] value)
+    {
+        return Int128.valueOf(
+                (((long) value[0]) << 32) | value[1],
+                (((long) value[2]) << 32) | value[3]);
+    }
+
+    public static Int128 valueOf(long[] value)
+    {
+        long high = value[0];
+        long low = value[1];
+        return valueOf(high, low);
+    }
+
+    public static Int128 valueOf(long high, long low)
+    {
+        return new Int128(high, low);
+    }
+
+    public static Int128 valueOf(String value)
+    {
+        return Int128.valueOf(new BigInteger(value));
+    }
+
+    public static Int128 valueOf(BigInteger value)
+    {
+        long low = value.longValue();
+        long high = 0;
+        try {
+            high = value.shiftRight(64).longValueExact();
+        }
+        catch (ArithmeticException e) {
+            throw new ArithmeticException("BigInteger out of Int128 range");
+        }
+
+        return new Int128(high, low);
+    }
+
+    public static Int128 valueOf(long value)
+    {
+        return new Int128(value >> 63, value);
+    }
+
+    public long getHigh()
+    {
+        return high;
+    }
+
+    public long getLow()
+    {
+        return low;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Int128 that = (Int128) o;
+        return high == that.high && low == that.low;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(high, low);
+    }
+
+    @Override
+    public int compareTo(Int128 other)
+    {
+        return compare(high, low, other.high, other.low);
+    }
+
+    @Override
+    public String toString()
+    {
+        return toBigInteger().toString();
+    }
+
+    public BigInteger toBigInteger()
+    {
+        return new BigInteger(toBigEndianBytes());
+    }
+
+    public byte[] toBigEndianBytes()
+    {
+        byte[] bytes = new byte[16];
+        toBigEndianBytes(bytes, 0);
+        return bytes;
+    }
+
+    public void toBigEndianBytes(byte[] bytes, int offset)
+    {
+        BIG_ENDIAN_LONG_VIEW.set(bytes, offset, high);
+        BIG_ENDIAN_LONG_VIEW.set(bytes, offset + Long.BYTES, low);
+    }
+
+    public long toLong()
+    {
+        return low;
+    }
+
+    public long toLongExact()
+    {
+        if (high != (low >> 63)) {
+            throw new ArithmeticException("Overflow");
+        }
+
+        return low;
+    }
+
+    public long[] toLongArray()
+    {
+        return new long[] {high, low};
+    }
+
+    public static int compare(long leftHigh, long leftLow, long rightHigh, long rightLow)
+    {
+        int comparison = Long.compare(leftHigh, rightHigh);
+        if (comparison == 0) {
```

Comment:
You mean like this?

```java
int high = Long.compare(leftHigh, rightHigh);
int low = Long.compareUnsigned(leftLow, rightLow);
return high == 0 ? low : high;
```

It's not clear that that's actually generates branchless code (need to check the assembly), and it has extra cos of computing low unconditionally if the VM is not smart enough to move the computation inside the `high == 0` check if it's not able to generate branchless code.

Reply:
Rather something like:
```
int highComparison = Long.compare(leftHigh, rightHigh);
int lowComparison = Long.compareUnsigned(leftLow, rightLow);
comparison = highComparison + ~(highComparison & 1) * lowComparison;
```
But I guess you can put it on the todo list and benchmark it later

Reply:
> `highComparison & 1`

That may not work. Long.compare is not guaranteed to return -1, 0 or 1. According to its javadocs, it will return "a value less than 0 if x < y", "a value greater than 0 if x > y".

It would need to be adjusted to something like:

```java
int high = Long.compare(aHigh, bHigh);
int low = Long.compareUnsigned(aLow, bLow);

int mask = ~(high | -high) >> 31;
int comparison = high + (mask & low);
```